File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/RoadRunner/messages/TwoDeadWheelInputsMessage.java
package org.firstinspires.ftc.teamcode.RoadRunner.messages;

import com.acmerobotics.roadrunner.ftc.PositionVelocityPair;

import org.firstinspires.ftc.robotcore.external.navigation.AngleUnit;
import org.firstinspires.ftc.robotcore.external.navigation.AngularVelocity;
import org.firstinspires.ftc.robotcore.external.navigation.YawPitchRollAngles;

public final class TwoDeadWheelInputsMessage {
    public long timestamp;
    public PositionVelocityPair par;
    public PositionVelocityPair perp;
    public double yaw;
    public double pitch;
    public double roll;
    public double xRotationRate;
    public double yRotationRate;
    public double zRotationRate;

    public TwoDeadWheelInputsMessage(PositionVelocityPair par, PositionVelocityPair perp, YawPitchRollAngles angles, AngularVelocity angularVelocity) {
        this.timestamp = System.nanoTime();
        this.par = par;
        this.perp = perp;
        {
            this.yaw = angles.getYaw(AngleUnit.RADIANS);
            this.pitch = angles.getPitch(AngleUnit.RADIANS);
            this.roll = angles.getRoll(AngleUnit.RADIANS);
        }
        {
            this.xRotationRate = angularVelocity.xRotationRate;
            this.yRotationRate = angularVelocity.yRotationRate;
            this.zRotationRate = angularVelocity.zRotationRate;
        }
    }
}

File: TeamCode/src/main/main/java/org/firstinspires/ftc/teamcode/drivetrain/messages/TwoDeadWheelInputsMessage.kt
package org.firstinspires.ftc.teamcode.drivetrain.messages

import com.acmerobotics.roadrunner.ftc.PositionVelocityPair
import org.firstinspires.ftc.robotcore.external.navigation.AngleUnit
import org.firstinspires.ftc.robotcore.external.navigation.AngularVelocity
import org.firstinspires.ftc.robotcore.external.navigation.YawPitchRollAngles

class TwoDeadWheelInputsMessage(
    par: PositionVelocityPair?,
    perp: PositionVelocityPair?,
    angles: YawPitchRollAngles,
    angularVelocity: AngularVelocity
) {
    var timestamp: Long
    var par: PositionVelocityPair?
    var perp: PositionVelocityPair?
    var yaw: Double = 0.0
    var pitch: Double = 0.0
    var roll: Double = 0.0
    var xRotationRate: Double = 0.0
    var yRotationRate: Double = 0.0
    var zRotationRate: Double = 0.0

    init {
        this.timestamp = System.nanoTime()
        this.par = par
        this.perp = perp
        run {
            this.yaw = angles.getYaw(AngleUnit.RADIANS)
            this.pitch = angles.getPitch(AngleUnit.RADIANS)
            this.roll = angles.getRoll(AngleUnit.RADIANS)
        }
        run {
            this.xRotationRate = angularVelocity.xRotationRate.toDouble()
            this.yRotationRate = angularVelocity.yRotationRate.toDouble()
            this.zRotationRate = angularVelocity.zRotationRate.toDouble()
        }
    }
}
